corelib/crayfish_renderer.cpp

```diff
@@ -297,264 +311,343 @@ void Renderer::drawVectorData(const Output* output)
+        float fixedShaftLength = cfg->ds.mFixedShaftLength;
+        xDist = cosAlpha * fixedShaftLength;
+        yDist = sinAlpha * fixedShaftLength;
+        break;
+    }
+    }
+
+    // Flip the Y axis (pixel vs real-world axis)
+    yDist *= -1.0;
+
+    if (qAbs(xDist) < 1 && qAbs(yDist) < 1)
+        return true;
+
+    // Determine the line coords
+    lineEnd = QPointF( lineStart.x() + xDist,
+                       lineStart.y() + yDist);
+
+    vectorLength = sqrt(xDist*xDist + yDist*yDist);
+
+    // Check to see if any of the coords are outside the QImage area, if so, skip the whole vector
+    if( lineStart.x() < 0 || lineStart.x() > cfg->outputSize.width() ||
+            lineStart.y() < 0 || lineStart.y() > cfg->outputSize.height() ||
+            lineEnd.x() < 0   || lineEnd.x() > cfg->outputSize.width() ||
+            lineEnd.y() < 0   || lineEnd.y() > cfg->outputSize.height() )
+        return true;
+
+    return false; //success
+}
+
+void Renderer::drawVectorDataStreamLines(QPainter& p) {
+    TraceRendererCache* cache = mMesh->getTraceCache();
+    for (int i=0; i<cache->getStreamLinesCount(); ++i) {
+        const TraceStreamLine& streamline = cache->getStreamLine(i);
+        for (int j=1; j<streamline.size(); ++j) {
+            const QPointF& startPoint = streamline.at(j-1);
+            const QPointF& endPoint = streamline.at(j);
+            p.drawLine(startPoint, endPoint);
+        }
     }
-  }
+}
+
+void Renderer::drawVectorDataAsTrace(QPainter& p) {
+    QVector<QColor> colors;
+    colors.append(QColor(Qt::transparent));
+
+    int steps = mCfg.ds.mVectorTraceAnimationSteps;
+    for (int i=0; i<steps; ++i) {
+        QColor c(mCfg.ds.mVectorColor);
+        c.setAlphaF(i/float(steps));
+        colors.append(c);
+    }
+
+
+    TraceRendererCache* cache = mMesh->getTraceCache();
+    int iter = cache->getNextIteration();
+
+    for (int i=0; i<cache->getStreamLinesCount(); ++i) {
+        const TraceStreamLine& streamline = cache->getStreamLine(i);
 
-  // Flip the Y axis (pixel vs real-world axis)
-  yDist *= -1.0;
+        for (int j=1; j<streamline.size(); ++j) {
+            const QPointF& startPoint = streamline.at(j-1);
+            const QPointF& endPoint = streamline.at(j);
 
-  if (qAbs(xDist) < 1 && qAbs(yDist) < 1)
-    return;
+            QLinearGradient gradient;
+            gradient.setStart(startPoint);
+            gradient.setFinalStop(endPoint);
+            gradient.setColorAt(0, colors.at((steps + j - iter - 1) % steps));
+            gradient.setColorAt(1, colors.at((steps + j - iter) % steps));
 
-  // Determine the line coords
-  QPointF lineEnd = QPointF( lineStart.x() + xDist,
-                             lineStart.y() + yDist);
+            QPen pen = p.pen();
+            pen.setBrush(QBrush(gradient));
+            p.setPen(pen);
 
-  float vectorLength = sqrt(xDist*xDist + yDist*yDist);
+            p.drawLine(startPoint, endPoint);
+        }
+    }
+}
```

Comment:
Originally I thought we would do the trace animation by using a cached image, and in every timestep fading out existing image content and painting into image just the latest bit of the path. But if this has reasonable performance, then we can keep it as it is...